File: skill-sync-backend/src/main/java/com/rutik/skill_sync_backend/chat/service/interfaces/ChatService.java
package com.rutik.skill_sync_backend.chat.service.interfaces;

import com.rutik.skill_sync_backend.chat.dto.request.ReadReceiptRequest;
import com.rutik.skill_sync_backend.chat.dto.request.SendMessageRequest;
import com.rutik.skill_sync_backend.chat.dto.response.MessageResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for Chat messaging operations.
 */
public interface ChatService {

    /**
     * Sends a new message in a conversation.
     *
     * @param currentUserId the ID of the user sending the message
     * @param request the send message request payload
     * @return the MessageResponse
     */
    MessageResponse sendMessage(Long currentUserId, SendMessageRequest request);

    /**
     * Retrieves messages for a conversation in a paginated manner.
     *
     * @param currentUserId the ID of the requesting user
     * @param conversationId the ID of the conversation
     * @param pageable pagination options
     * @return page of MessageResponse
     */
    Page<MessageResponse> getMessages(Long currentUserId, Long conversationId, Pageable pageable);

    /**
     * Marks a specific message and all prior messages in the conversation as read.
     *
     * @param currentUserId the ID of the user marking messages as read
     * @param request the read receipt request payload
     */
    void markAsRead(Long currentUserId, ReadReceiptRequest request);

    /**
     * Get the total unread messages count for the user across all conversations.
     *
     * @param currentUserId the ID of the user
     * @return the total unread messages count
     */
    long getUnreadCount(Long currentUserId);
}

File: frontend/src/services/chatService.js
import api from './api';

/**
 * Service for Chat REST operations.
 */
const chatService = {
  /**
   * Retrieves paginated conversation summaries for the current user.
   */
  getConversations: async (page = 0, size = 20) => {
    const response = await api.get('/chat/conversations', {
      params: { page, size }
    });
    return response.data;
  },

  /**
   * Retrieves full details of a specific conversation (participants, sessions timeline, initial messages).
   */
  getConversationDetails: async (conversationId) => {
    const response = await api.get(`/chat/conversations/${conversationId}`);
    return response.data;
  },

  /**
   * Retrieves a paginated slice of messages for a specific conversation.
   */
  getMessages: async (conversationId, page = 0, size = 20) => {
    const response = await api.get(`/chat/conversations/${conversationId}/messages`, {
      params: { page, size }
    });
    return response.data;
  },

  /**
   * Creates a new conversation with the given recipient.
   */
  createConversation: async (recipientId) => {
    const response = await api.post('/chat/conversations', { recipientId });
    return response.data;
  },

  /**
   * Sends a message via REST API.
   */
  sendMessage: async (conversationId, sessionId, content, type = 'TEXT', clientMessageId = null) => {
    const response = await api.post('/chat/messages', {
      conversationId,
      sessionId,
      content,
      type,
      clientMessageId
    });
    return response.data;
  },

  /**
   * Marks a conversation's messages up to the given message ID as read.
   */
  markAsRead: async (conversationId, messageId) => {
    const response = await api.put('/chat/messages/read', {
      conversationId,
      messageId
    });
    return response.data;
  },

  /**
   * Retrieves total unread message count across all conversations.
   */
  getUnreadCount: async () => {
    const response = await api.get('/chat/unread-count');
    return response.data;
  }
};

export default chatService;
